### PR TITLE
lnwallet: Move htlc mutation out of evaluate htlc view and expose resolutions

### PR DESCRIPTION
**Pull request by carlaKC**
This PR moves mutation of htlcs previously in `evaluateHtlcView` further up the call stack to `computeView` and exposes a set of resolutions resulting from the new view. It partially addresses #4088 by exposing the htlcs that are resolved as of the view evaluation we have performed. Future changes will bubble this resolution set up further for atomic persistence of htlc settles.

### CODE REVIEW

File: lnwallet/channel.go [hunk 1]
```diff
@@ -2617,6 +2569,64 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
 	return newView, nil
 }
 
+// fetchParent is a function which is used to look up an update's parent in
+// the appropriate update log.
+type fetchParent func(entry *PaymentDescriptor,
+	remoteLog bool) (*PaymentDescriptor, error)
+
+// getFetchParent is a helper that returns a function which returns a function
```
Comment (halseth) [inline lnwallet/channel.go]: 😵

File: lnwallet/channel.go [hunk 2]
```diff
@@ -2617,6 +2569,64 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
...
+// for looking up parent entries in the appropriate log.
+func (lc *LightningChannel) getFetchParent(remoteChain bool) fetchParent {
+	return func(entry *PaymentDescriptor,
```
Comment (halseth) [inline lnwallet/channel.go]: could `getFetchParent` rather be this returned method?

File: lnwallet/channel_test.go [hunk 3]
```diff
@@ -7306,3 +7305,385 @@ func TestChannelMaxFeeRate(t *testing.T) {
 	assertMaxFeeRate(0.000001, 690)
 	assertMaxFeeRate(0.0000001, chainfee.FeePerKwFloor)
 }
+
+// TestEvaluateView tests the creation of a htlc view and the opt in mutation of
+// send and receive balances. This test does not check htlc mutation on a htlc
+// level.
+func TestEvaluateView(t *testing.T) {
```
Comment (halseth) [inline lnwallet/channel_test.go]: awesome, about time! 💯

Comment (halseth) [reply lnwallet/channel_test.go]: (skipping review of actual tests for now)

File: lnwallet/channel.go [hunk 4]
```diff
@@ -2527,9 +2558,10 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
 		// and it hasn't been processed, yet, the increment our state
 		// tracking the total number of satoshis we've sent within the
 		// channel.
-		if mutateState && entry.EntryType == Settle && !remoteChain &&
+		if entry.EntryType == Settle && !remoteChain &&
 			entry.removeCommitHeightLocal == 0 {
-			lc.channelState.TotalMSatSent += entry.Amount
+
+			evaluation.sentTotal += entry.Amount
```
Comment (halseth) [inline lnwallet/channel.go]: I like where this is heading! Could the same be done for the HTLC mutation? That being everywhere we now check `mutateState`, add the mutation to the `viewEvaluation` (could even be a list of `func`'s to run when we mutate), then evaluate this outside `evaluateView`. Feels like that would significantly reduce the diff ...

File: lnwallet/channel.go [hunk 5]
```diff
@@ -2440,39 +2440,287 @@ func fundingTxIn(chanState *channeldb.OpenChannel) wire.TxIn {
...
 // producing a final view which is the result of properly applying all adds,
 // settles, timeouts and fee updates found in both logs. The resulting view
 // returned reflects the current state of HTLCs within the remote or local
-// commitment chain, and the current commitment fee rate.
-//
-// If mutateState is set to true, then the add height of all added HTLCs
-// will be set to nextHeight, and the remove height of all removed HTLCs
-// will be set to nextHeight. This should therefore only be set to true
-// once for each height, and only in concert with signing a new commitment.
-// TODO(halseth): return htlcs to mutate instead of mutating inside
```
Comment (halseth) [inline lnwallet/channel.go]: nice to finally get rid of this 😅

**Comment by carlaKC** [comment]
The original reason for this PR was to obtain a list of settles further up the call stack, which isn't actually needed (we have sufficient data already). Since this is sensitive code to touch, waiting on a peep from @Roasbeef to decide whether we want to go forward with this refactor for the sake of refactoring. If not, will strip the tests out so we still get the coverage increase.

**Comment by carlaKC** [comment]
Closing for now, can always reopen if we choose to go ahead.

**Pull request closed without merging**